File: bitcamp-web-03/src/main/java/bitcamp/pms/dao/MemberDao.java
package bitcamp.pms.dao;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.List;

import bitcamp.pms.domain.Member;

public class MemberDao {
    static {
        try {
            Class.forName("com.mysql.jdbc.Driver");
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    
     String jdbcUrl;
     String username;
     String password;
    
    public MemberDao(
            String jdbcUrl, String username, String password) {
        this.jdbcUrl = jdbcUrl;
        this.username = username;
        this.password = password;
    }
    
    
    public List<Member> selectList() throws Exception {
        try (
             Connection con = DriverManager.getConnection(
                        jdbcUrl,username, password);
            PreparedStatement stmt = con.prepareStatement(
                "select mid, email from pms2_member");
            ResultSet rs = stmt.executeQuery();) {
            
            ArrayList<Member> list = new ArrayList<>();
            while (rs.next()) {
                
                Member member = new Member();
                member.setId(rs.getString("mid"));// 메모리를 다루는 연산자    
                member.setEmail(rs.getString("email"));
                
                list.add(member);
     
            }
            return list;
        }
    }
    
    public Member selectOne(String id) throws Exception{
        try (
              Connection con = DriverManager.getConnection(
                      jdbcUrl,username, password);
                PreparedStatement stmt = con.prepareStatement(
                    "select mid,email from pms2_member where mid=?");) {
                
                stmt.setString(1, id);
                
                try (ResultSet rs = stmt.executeQuery();) {
                        if (!rs.next()) 
                            return null;
                        
                        Member member = new Member();
                        member.setId(id);
                        member.setEmail(rs.getString("email"));
                        return member;
                    }
                    
        }
    }
  public int update(Member member) throws Exception {
        
        try (
            Connection con = DriverManager.getConnection(
                        jdbcUrl,username, password);
            PreparedStatement stmt = con.prepareStatement(
                "update pms2_member set email=?, pwd=sha2(?,224) where mid=?");) {
            
            stmt.setString(1, member.getEmail());
            stmt.setString(2, member.getPassword());
            stmt.setString(3, member.getId());
            return stmt.executeUpdate();
        
        }
    }
  public int delete(String id) throws Exception {
      try (
          Connection con = DriverManager.getConnection(
                      jdbcUrl,username, password);
          PreparedStatement stmt = con.prepareStatement(
              "delete from pms2_member where mid=?");
             ) {
          System.out.println("삭제가 되나??");
          stmt.setString(1, id);
          
      
      return stmt.executeUpdate();
  }
  }
  public int insert(Member member) throws Exception {
      try (
          Connection con = DriverManager.getConnection(
                      jdbcUrl,username, password);
          PreparedStatement stmt = con.prepareStatement(
              "insert into pms2_member(mid,email,pwd) values(?,?,?)");) {
          
          stmt.setString(1, member.getId());
          stmt.setString(2, member.getEmail());
          stmt.setString(3, member.getPassword());
      
          return stmt.executeUpdate();
      }
  }
  
}

File: bitcamp-nodejs/src/step03/memberdao.js
// 주제:  dao 모듈 만들기
//
//
var pool;
exports.setConnectionPool=(connectionPool)=>{
  pool = connectionPool;
};
exports.list = (pageNo=1,pageSize=3,handler)=>{
  var startIndex = (pageNo - 1) * pageSize;

  //this 생략 불가
  pool.query('select mid, email from pms2_member limit ?, ?',
          [startIndex, pageSize],
          function(err, results) {
        handler(err,results);
  });
}

exports.add = (data,handler)=>{
  pool.query(
          'insert into pms2_member(mid,email,pwd)\
          values(?, ?, password(?))',
      [req.query.id, req.query.email, req.query.password],
      function(err, result) {
          handler(err,result)

  });
};

exports.update = (data,handler)=>{
    pool.query(
            'update pms2_member set\
             email=?,\
             pwd=?\
             where mid=?',
        [req.query.email,
         req.query.password,
         req.query.id],
        function(err, result) {
          handler(err,result);
        });
      };
exports.remove = (data,handler)=>{
  pool.query('delete from pms2_member where mid=?',
      [req.query.id],
      function(err, result) {
        handler(err,result)
      });
    };
